pjsip/src/pjsua-lib/pjsua_media.c

```diff
@@ -3983,7 +3983,9 @@ pj_status_t pjsua_media_channel_update(pjsua_call_id call_id,
 	    }
 
 	    /* Check if no media is active */
-	    if (local_sdp->media[mi]->desc.port == 0) {
+	    if ((local_sdp->media[mi]->desc.port == 0) || 
+	        (si->dir == PJMEDIA_DIR_NONE))	    
+	    {
```

Comment:
This seems to introduce rather major behavioral change. Note that port zero and a=inactive in SDP are not really the same. When the port is zero, media is completely disabled (no media transport is created), otherwise it still needs to have media transport, e.g: RTP keep alive may still active, RTCP too.

Reply:
In the above case (`a=inactive`), when `on_call_media_state()` is called the media state will be set to `PJSUA_CALL_MEDIA_ACTIVE`. So, app needs to check for the call media dir in addition to the media state.
e.g: `media dir = PJMEDIA_DIR_NONE` indicates `a=inactive`.

Reply:
> In the above case (`a=inactive`), when `on_call_media_state()` is called the media state will be set to `PJSUA_CALL_MEDIA_ACTIVE`. So, app needs to check for the call media dir in addition to the media state. e.g: `media dir = PJMEDIA_DIR_NONE` indicates `a=inactive`.

Is this the current behavior or the new behavior after the patch?

Anyway, my concern is that media transports should remain active (for keep-alive & RTCP) on `a=inactive` scenario, not destroyed as when SDP port is zero.

Reply:
I'm pointing out the current behavior (call has `a=inactive` scenario).
Users usually only checks for media state when call is connected to perform conf connect.
In the case of `a=inactive`, the conf connect is not as expected although the media is active.
Checking for media dir will also be required besides the media state check.

Reply:
Okay. So as mentioned before, my concern on this PR patch, will the media transport now destroyed when `a=inactive`?

Reply:
With this patch the media transport (RTP/RTCP) will still be created, not destroyed however it will not be started. 
For `keep-alive` it will not change the behavior. It is part of the stream, and in this case the stream is not created.

Another alternative would be to set the media state to PJSUA_CALL_MEDIA_NONE:
```
if (call->local_hold)
    call_med->state = PJSUA_CALL_MEDIA_LOCAL_HOLD;
else if (call_med->dir == PJMEDIA_DIR_DECODING)
    call_med->state = PJSUA_CALL_MEDIA_REMOTE_HOLD;
else if (call_med->dir == PJMEDIA_DIR_NONE)
    call_med->state = PJSUA_CALL_MEDIA_NONE;
else
    call_med->state = PJSUA_CALL_MEDIA_ACTIVE;
```

Reply:
The above alternative looks much safer, i.e: only affects the call media state value, right? If it does, I'll vote for this approach :)

Reply:
Yes, it should only affects the call media state.

Reply:
What about the double hold scenario?
Do we want the call media state to be `PJSUA_CALL_MEDIA_LOCAL_HOLD` or `PJSUA_CALL_MEDIA_NONE`?

Reply:
I think we should preserve the existing state (keep it `PJSUA_CALL_MEDIA_LOCAL_HOLD`).
It would prevent app breaking which uses that state.

Reply:
Adding to the stream creation. The stream will be created if media changes is detected. e.g: on the case of call hold, media changes from `sendrecv` -> `sendonly`. The stream will be recreated. However on initial call if the SDP contains `inactive`, then stream will not be created since no media change occurs ([ref](https://github.com/pjsip/pjproject/blob/master/pjsip/src/pjsua-lib/pjsua_media.c#L4069)). 
I agree that the original patch will introduce major behavior change. So it would be safe to discard this solution.

alternative 1:
```
if (call->local_hold)
    call_med->state = PJSUA_CALL_MEDIA_LOCAL_HOLD;
else if (call_med->dir == PJMEDIA_DIR_DECODING)
    call_med->state = PJSUA_CALL_MEDIA_REMOTE_HOLD;
else if (call_med->dir == PJMEDIA_DIR_NONE)
    call_med->state = PJSUA_CALL_MEDIA_NONE;
else
    call_med->state = PJSUA_CALL_MEDIA_ACTIVE;
```

alternative 2:
```
if (call->local_hold)
    call_med->state = PJSUA_CALL_MEDIA_LOCAL_HOLD;
else if (call_med->dir == PJMEDIA_DIR_DECODING)
    call_med->state = PJSUA_CALL_MEDIA_REMOTE_HOLD;
else if (call_med->dir== PJMEDIA_DIR_NONE)
    /* If remote sends `a=inactive`, we'll treat it as a remote hold */
    call_med->state =  is_answerer? PJSUA_CALL_MEDIA_REMOTE_HOLD : PJSUA_CALL_MEDIA_NONE;
    /* ==== end of new lines ====*/
else
    call_med->state = PJSUA_CALL_MEDIA_ACTIVE;
```

About the purpose patch.
scenario 1: (call hold - no issue)
caller sends `sendonly`
callee sends `recvonly`

| Patch | Caller | Callee |
| ----------- | ----------- | ----------- |
| No patch     |  PJSUA_CALL_MEDIA_LOCAL_HOLD   | PJSUA_CALL_REMOTE_HOLD |
| alt 1     |  PJSUA_CALL_MEDIA_LOCAL_HOLD   | PJSUA_CALL_REMOTE_HOLD |
| alt 2     |  PJSUA_CALL_MEDIA_LOCAL_HOLD   | PJSUA_CALL_REMOTE_HOLD |

scenario 2: 
caller sends `sendrecv`
callee sends `inactive`

| Patch | Caller | Callee |
| ----------- | ----------- | ----------- |
| No patch     |  PJSUA_CALL_MEDIA_ACTIVE  | PJSUA_CALL_MEDIA_ACTIVE  |
| alt 1     |  PJSUA_CALL_MEDIA_NONE | PJSUA_CALL_MEDIA_NONE |
| alt 2     |  PJSUA_CALL_MEDIA_NONE   | PJSUA_CALL_MEDIA_NONE ? |

scenario 3: 
caller sends `sendonly`
callee sends `inactive`

| Patch | Caller | Callee |
| ----------- | ----------- | ----------- |
| No patch     |  PJSUA_CALL_MEDIA_LOCAL_HOLD   | PJSUA_CALL_MEDIA_ACTIVE  |
| alt 1     |  PJSUA_CALL_MEDIA_LOCAL_HOLD | PJSUA_CALL_MEDIA_NONE |
| alt 2     |  PJSUA_CALL_MEDIA_LOCAL_HOLD | PJSUA_CALL_REMOTE_HOLD |

For alt 2, we need to check if the INVITE message contains `sendonly` or `sendrecv`.

Reply:
For alternative 2:
If we set media state based on remote SDP has `a=inactive`, then we should also handle if local is the one that sent `a=inactive`. e.g: should the offerer set to PJSUA_CALL_MEDIA_LOCAL_HOLD?
Because of this, I would prefer the alt 1.

Reply:
The idea of alt 2 is based on the assumption of that some implementations (not PJSUA, such as in the original report) send `inactive` (instead of `sendonly`) when initiating call hold. That's why the alt 2 only checks if we are the answerer. It is just an attempt to handle such implementations. It of course can be improved to handle other scenario.

Note that PJSUA by default (note that app may override the SDP via PJSUA callback) will never send `inactive` for initiating call hold (unless remote initates call hold too).

Reply:
Btw, PJSUA now actually has the API/capability to set media direction (without having to modify the SDP), i.e. using `PJSUA_CALL_SET_MEDIA_DIR` flag, for example:
```
    call_opt.media_dir[0] = PJMEDIA_DIR_NONE;
    call_opt.flag = PJSUA_CALL_SET_MEDIA_DIR;
    pjsua_call_reinvite2(...,&call_opt,...);
```
So we probably need to take this into consideration as well.

Reply:
Right, the `PJSUA_CALL_SET_MEDIA_DIR` needs to be handled as well. e.g: for call hold, the `local_hold` field will only be set when we call `pjsua_call_set_hold()/pjsua_call_set_hold2()`. We can set the local_hold field on pjsua_call_make_call()/pjsua_call_reinvite()/pjsua_call_update() when `PJSUA_CALL_SET_MEDIA_DIR + PJMEDIA_DIR_ENCODING` is set.

Back to alternative 2. If we are going to set `PJSUA_CALL_MEDIA_REMOTE_HOLD` when remote sent `a=inactive`, then 
it would be expected that local would be set to `PJSUA_CALL_MEDIA_LOCAL_HOLD`. This means the `PJSUA_CALL_MEDIA_NONE` will no longer be used. Since the `a=inactive` can be issued by the offerer/answerer.
```
if (call->local_hold)
    call_med->state = PJSUA_CALL_MEDIA_LOCAL_HOLD;
else if (call_med->dir == PJMEDIA_DIR_DECODING)
    call_med->state = PJSUA_CALL_MEDIA_REMOTE_HOLD;
else if (call_med->dir== PJMEDIA_DIR_NONE)
    /* If remote sends `a=inactive`, we'll treat it as a remote hold */
    call_med->state =  is_answerer? PJSUA_CALL_MEDIA_REMOTE_HOLD : PJSUA_CALL_MEDIA_LOCAL_HOLD;
    /* ==== end of new lines ====*/
else
    call_med->state = PJSUA_CALL_MEDIA_ACTIVE;
```
`PJSUA_CALL_MEDIA_NONE` is not used anymore and might break app checking for such state.

For the original report, the issue with call hold is caused by the original patch
```
if ((local_sdp->media[mi]->desc.port == 0) || 
	        (si->dir == PJMEDIA_DIR_NONE))	    
{
		/* Update call media state and direction */
		call_med->state = PJSUA_CALL_MEDIA_NONE;
		call_med->dir = PJMEDIA_DIR_NONE;
}
```
This will break the current call hold scenario. e.g: app expects `PJSUA_CALL_LOCAL_HOLD`, but `PJSUA_CALL_MEDIA_NONE`.
It didn't expect to receive new state e.g: `PJSUA_CALL_REMOTE_HOLD` when receiving `a=inactive`.